core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalH2OBackend.scala

```diff
@@ -62,78 +63,105 @@ class InternalH2OBackend(@transient val hc: H2OContext) extends SparklingBackend
+
+    H2OStarter.start(launcherArgs, false)
+    NodeDesc(SparkEnv.get.executorId, H2O.SELF_ADDRESS.getHostName, H2O.API_PORT)
+  }
+
 
-    // H2O is already started on the nodes, make sure that we kill the cloud when a new executor joins
-    if (!(hc.sparkContext.isLocal || hc.sparkContext.master.startsWith("local-cluster[")) && hc.getConf.isClusterTopologyListenerEnabled) {
+  def startH2OWorker(conf: H2OConf): NodeDesc = {
+    var args = InternalBackendUtils.getH2OWorkerArgs(conf)
+    val launcherArgs = InternalBackendUtils.toH2OArgs(args)
+
+    H2OStarter.start(launcherArgs, true)
+    NodeDesc(SparkEnv.get.executorId, H2O.SELF_ADDRESS.getHostName, H2O.API_PORT)
+  }
+
+  private def startH2OClient(conf: H2OConf, nodes: Array[NodeDesc]): NodeDesc = {
+    val args = InternalBackendUtils.getH2OClientArgs(conf)
+    val launcherArgs = InternalBackendUtils.toH2OArgs(args, nodes)
+
+    H2OStarter.start(launcherArgs, false)
+    NodeDesc(SparkEnv.get.executorId, H2O.SELF_ADDRESS.getHostAddress, H2O.API_PORT)
+  }
+
+  private def registerNewExecutorListener(hc: H2OContext): Unit = {
+    if (!hc.sparkContext.master.startsWith("local-cluster[") && hc.getConf.isClusterTopologyListenerEnabled) {
       hc.sparkContext.addSparkListener(new SparkListener {
         override def onExecutorAdded(executorAdded: SparkListenerExecutorAdded): Unit = {
           log.warn("New spark executor joined the cloud, however it won't be used for the H2O computations.")
         }
       })
     }
+  }
+
+  private def tearDownEndpoints(endpoints: Array[RpcEndpointRef]): Unit = endpoints.foreach(_.send(StopEndpointMsg))
 
-    // Connect to a cluster via H2O client, but only in non-local case
-    if (!hc.sparkContext.isLocal) {
-      logTrace("Sparkling H2O - DISTRIBUTED mode: Waiting for " + executors.length)
-      // Get arguments for this launch including flatfile ( Do not use IP if network mask is specified)
-      val h2oClientArgs = InternalBackendUtils.toH2OArgs(InternalBackendUtils.getH2OClientArgs(hc.getConf), hc.getConf, executors)
-      logInfo(s"Starting H2O client on the Spark Driver (${getHostname(SparkEnv.get)}): ${h2oClientArgs.mkString(" ")}")
-      // Launch H2O
-      H2OStarter.start(h2oClientArgs, false)
+  private def registerEndpoints(hc: H2OContext): Array[RpcEndpointRef] = {
+    val endpoints = new SpreadRDDBuilder(hc, InternalBackendUtils.guessTotalExecutorSize(hc.sparkContext)).build()
+    endpoints.map { ref =>
+      SparkEnv.get.rpcEnv.setupEndpointRef(ref.address, ref.name)
     }
-    // And wait for right cluster size
-    H2O.waitForCloudSize(executors.length, hc.getConf.cloudTimeout)
-    // Register web API for client
-    RestAPIManager(hc).registerAll()
-    H2O.startServingRestApi()
-    executors
   }
 
-  override def epilog = ""
-}
+  private def startH2OWorkers(endpoints: Array[RpcEndpointRef], conf: H2OConf): Array[NodeDesc] = {
+    endpoints.map {
+      _.askSync[NodeDesc](StartH2OWorkersMsg(conf))
```

Comment:
Would it be nice to handle also situations when the start is not successful.

Reply:
Definitely good idea https://0xdata.atlassian.net/browse/SW-1302